### PR TITLE
fix(relay): flush relayed connection once idle

### PR DESCRIPTION
**Pull request by mxinden**
## Description

As a relay, when forwarding data between relay-connection-source and -destination and vice versa, flush write side when read currently has no more data available.

## Notes & open questions

Credit goes to @MarcoPolo for discovering this!

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/relay/src/copy_future.rs [hunk 1]
```diff
@@ -150,95 +157,59 @@ fn forward_data<S: AsyncBufRead + Unpin, D: AsyncWrite + Unpin>(
...
     use futures::executor::block_on;
-    use futures::io::{AsyncRead, AsyncWrite};
+    use futures::io::{AsyncRead, AsyncWrite, BufReader, BufWriter};
     use quickcheck::QuickCheck;
     use std::io::ErrorKind;
     use std::pin::Pin;
     use std::task::{Context, Poll};
     use std::time::Duration;
 
-    struct Connection {
```
Comment (mxinden) [inline protocols/relay/src/copy_future.rs:162]: To make confusing diff easier to read:

- `Connection` is only used in `quickckeck` test, thus moved into `quickcheck` test.
- `PendingConnection` is only used in `max_circuit_duration` test, thus moved into `max_circuit_duration` test.

File: protocols/relay/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+## 0.15.2 - unreleased
+
+- As a relay, when forwarding data between relay-connection-source and -destination and vice versa, flush write side when read currently has no more data available.
+  See [PR XXX].
+
+[PR XXX]: https://github.com/libp2p/rust-libp2p/pull/XXX
```
Comment (thomaseizinger) [inline protocols/relay/CHANGELOG.md]:
```suggestion
- As a relay, when forwarding data between relay-connection-source and -destination and vice versa, flush write side when read currently has no more data available.
  See [PR 3765].

[PR 3765]: https://github.com/libp2p/rust-libp2p/pull/3765
```

File: protocols/relay/src/copy_future.rs [hunk 3]
```diff
@@ -150,95 +157,59 @@ fn forward_data<S: AsyncBufRead + Unpin, D: AsyncWrite + Unpin>(
 
 #[cfg(test)]
 mod tests {
-    use super::CopyFuture;
+    use super::{forward_data, CopyFuture};
```
Comment (thomaseizinger) [inline protocols/relay/src/copy_future.rs]:
```suggestion
    use super::*;
```

File: protocols/relay/src/copy_future.rs [hunk 4]
```diff
@@ -288,4 +295,124 @@ mod tests {
             block_on(copy_future).expect_err("Expect maximum circuit duration to be reached.");
         assert_eq!(error.kind(), ErrorKind::TimedOut);
     }
+
+    #[test]
+    fn forward_data_should_flush_on_pending_source() {
+        struct NeverEndingSource {
```
Comment (thomaseizinger) [inline protocols/relay/src/copy_future.rs:301]:
```suggestion
        struct NeverEndingStory {
```

File: protocols/relay/src/copy_future.rs [hunk 5]
```diff
@@ -288,4 +295,124 @@ mod tests {
...
+                forward_data(&mut source, &mut destination, &mut cx),
+                Poll::Ready(Ok(1)),
+            ),
+            "Expect `forward_data` to forward one read from the source to the wrapped destination."
+        );
+        assert_eq!(
+            destination.get_ref().method_calls.as_slice(), &[],
+            "Given that destination is wrapped with a `BufWrite`, the write doesn't (yet) make it to \
+            the destination. The source might have more data available, thus `forward_data` has not \
+            yet flushed.",
```
Comment (thomaseizinger) [inline protocols/relay/src/copy_future.rs:385]:
```suggestion
            "Expect BufWriter to still have space and thus not forward the poll_write call",
```

File: protocols/relay/src/copy_future.rs [hunk 6]
```diff
@@ -288,4 +295,124 @@ mod tests {
...
+            ) -> Poll<std::io::Result<()>> {
+                self.method_calls.push(Method::PollClose);
+                Poll::Ready(Ok(()))
+            }
+        }
+
+        // The source has two reads available, handing them out on `AsyncRead::poll_read` one by one.
+        let mut source = BufReader::new(NeverEndingSource { read: vec![1, 2] });
+
+        // The destination is wrapped by a `BufWrite` with a capacity of `3`, i.e. one larger than
```
Comment (thomaseizinger) [inline protocols/relay/src/copy_future.rs]:
```suggestion
        // The destination is wrapped by a `BufWriter` with a capacity of `3`, i.e. one larger than
```

**Review by thomaseizinger — APPROVED** [review]

**Comment by mxinden** [comment]
`quic-v1` failure:

```
2023-04-11T11:36:48.2299326Z     'rust-libp2p-head_x_rust-v0_51_0__quic-v1_-dialer-1    | Error: Connection refused (os error 111)\n' +
```
https://pipelines.actions.githubusercontent.com/serviceHosts/dceb4490-106f-45d6-a769-717b2af3d905/_apis/pipelines/1/runs/19034/signedlogcontent/2?urlExpires=2023-04-11T12%3A01%3A48.2772170Z&urlSigningMethod=HMACV1&urlSignature=fJLXS9uv%2BzG9b1ejGHxicCSUdxSr4oEKVBKP0dxNMmc%3D

**Comment by mxinden** [comment]
//CC @p-shahi for visibility for the universal connectivity project.

**Pull request merged**